kusto/kcsb.go

```diff
@@ -0,0 +1,362 @@
+	kcsb.ApplicationKey = appKey
+	kcsb.AuthorityId = authorityID
+	return kcsb
+}
+
+//Creates a KustoConnection string builder that will authenticate with AAD application using a certificate.
+func (kcsb connectionStringBuilder) WithAppCertificate(appId string, certificate string, thumprint string, sendCertChain bool, authorityID string) connectionStringBuilder {
+	assertIfEmpty(dataSource, kcsb.DataSource)
+	assertIfEmpty(applicationCertificate, certificate)
+	assertIfEmpty(applicationCertificateThumbprint, thumprint)
+	assertIfEmpty(authorityId, authorityID)
+	kcsb.ApplicationClientId = appId
+	kcsb.AuthorityId = authorityID
+
+	kcsb.ApplicationCertificate = certificate
+	kcsb.ApplicationCertificateThumbprint = thumprint
+	kcsb.SendCertificateChain = sendCertChain
+	return kcsb
+}
+
+// Creates a KustoConnection string builder that will authenticate with AAD application and an application token.
+func (kcsb connectionStringBuilder) WithApplicationToken(appId string, appToken string) connectionStringBuilder {
+	assertIfEmpty(dataSource, kcsb.DataSource)
+	assertIfEmpty(applicationToken, appToken)
+	kcsb.ApplicationToken = appToken
+	return kcsb
+}
+
+// Creates a KustoConnection string builder that will use existing authenticated az cli profile password.
+func (kcsb connectionStringBuilder) WithAzCli() connectionStringBuilder {
+	assertIfEmpty(dataSource, kcsb.DataSource)
+	kcsb.AZCLI = true
+	return kcsb
+}
+
+/*
+Creates a KustoConnection string builder that will authenticate with AAD application, using
+  an application token obtained from a Microsoft Service Identity endpoint. An optional user
+  assigned application ID can be added to the token.
+*/
+func (kcsb connectionStringBuilder) WithManagedServiceID(clientID string, resId string) connectionStringBuilder {
+	assertIfEmpty(dataSource, kcsb.DataSource)
+	kcsb.MSIAuthentication = true
+	if !isEmpty(clientID) {
+		kcsb.ManagedServiceIdentity = clientID
+	} else if !isEmpty(resId) {
+		kcsb.ManagedServiceIdentity = resId
+	}
+	return kcsb
+}
+
+func (kcsb connectionStringBuilder) WithInteractiveLogin(clientID string, authorityID string, redirectURL string) connectionStringBuilder {
+	assertIfEmpty(dataSource, kcsb.DataSource)
+	if !isEmpty(clientID) {
+		kcsb.ApplicationClientId = clientID
+	}
+	if !isEmpty(authorityID) {
+		kcsb.AuthorityId = authorityID
+	}
+	if !isEmpty(redirectURL) {
+		kcsb.RedirectURL = redirectURL
+	}
+	kcsb.InteractiveLogin = true
+	return kcsb
+}
+
+func (kcsb connectionStringBuilder) AttachClientOptions(options *azcore.ClientOptions) connectionStringBuilder {
+	assertIfEmpty(dataSource, kcsb.DataSource)
+	if options == nil {
+		kcsb.ClientOptions = options
+	}
+	return kcsb
+}
+
+// Method to be used for generating TokenCredential
+func (kcsb *connectionStringBuilder) getTokenProvider(ctx context.Context) (*tokenProvider, error) {
+	tkp := &tokenProvider{}
+	//Fetches cloud meta data
+	fetchedCI, cierr := GetMetadata(context.Background(), kcsb.DataSource)
```

Comment:
Having the metadata fetch here is something we try to avoid in the other sdks.
Because it means that just creating a client is a blocking operation that waits on a web request, plus we don't need to do it for all providers - like the constant token one